client/transport/cancellable/cancellable.go

```diff
@@ -91,23 +92,24 @@ func Do(ctx context.Context, client transport.Sender, req *http.Request) (*http.
 // Close is called or a Read fails on the underlying ReadCloser.
 type notifyingReader struct {
 	io.ReadCloser
-	notify chan<- struct{}
+	notify     chan<- struct{}
+	notifyOnce sync.Once
 }
 
 func (r *notifyingReader) Read(p []byte) (int, error) {
 	n, err := r.ReadCloser.Read(p)
-	if err != nil && r.notify != nil {
-		close(r.notify)
-		r.notify = nil
+	if err != nil {
+		r.notifyOnce.Do(func() {
+			close(r.notify)
+		})
 	}
 	return n, err
 }
 
 func (r *notifyingReader) Close() error {
 	err := r.ReadCloser.Close()
-	if r.notify != nil {
```

Comment:
Shouldn't `r.notify` should still be check for `nil`? There's no warranty that the field was assigned to, is there?